Tabular dev

stewartboyd119 opened the pull request:
Added a feature "tabular" for confirm a list of points (x,y) are within tolerance in abaverify

acbergan commented on abaverify/main.py:
```diff
@@ -452,7 +452,21 @@ def _runAssertionsOnResults(self, jobName):
+                        reference_val = r['referenceValue'][i]
+                        delta = r['tolerance'][i]
+                        if isinstance(computed_val, tuple):
+                            # --- when there exists a tuple as a reference val then all other results and deltas
```
Why the '---' in the comment?

stewartboyd119 replied:
I've used this the form # --- in the past for comments. The idea being two fold. One its eye catching when there's no syntax highlighting. Two it differentiates a comment between commented out code. It's just a habit. I see that it's not inline with the rest of the project. Will update

acbergan commented on abaverify/main.py:
```diff
@@ -452,7 +452,21 @@ def _runAssertionsOnResults(self, jobName):
+                        computed_val = r['computedValue'][i]
+                        reference_val = r['referenceValue'][i]
+                        delta = r['tolerance'][i]
+                        if isinstance(computed_val, tuple):
```
Does it make more sense to check the reference value since the computed value could be garbage in the case that the test failed?

stewartboyd119 replied:
agreed

acbergan commented on abaverify/main.py:
```diff
@@ -452,7 +452,21 @@ def _runAssertionsOnResults(self, jobName):
+                            # --- should also be tuples
+                            self.assertEqual(len(computed_val), len(reference_val),
+                                             "Specified reference value should be same length as Computed value")
+                            self.assertEqual(len(computed_val), len(delta),
```
I would like to see the tolerance have the option to be a single value that's applied to all computed y's (seems like a more likely use-case to me). If you want to keep the option to specify the tolerance as a list of tuples as it is currently, that's fine with me. Also, please refactor 'delta' to 'tolerance' so that line 469 reads: self.assertAlmostEqual(computed_val, reference_val, delta=tolerance) 

or similar. I'm suggesting this to keep the naming consistent.

stewartboyd119 replied:
Newest commit allows both. If the tolerance is specified as a single tuple it as the tolerance used against all computed vals. Otherwise the corresponding tolerance (by common index) is used

acbergan commented on abaverify/main.py:
```diff
@@ -452,7 +452,21 @@ def _runAssertionsOnResults(self, jobName):
+                                             "Specified reference value should be same length as Computed value")
+                            self.assertEqual(len(computed_val), len(delta),
+                                             "Specified delta should be same length as Computed value")
+                            # --- loop through entries in tuple
```
Lines 465-467 are nice and clean!

acbergan commented on abaverify/processresults.py:
```diff
@@ -45,6 +45,9 @@
     Last output value.
+8. tabular
+    Given a list of (x,y) points confirms that a point, within the specified, tolerance exists in post processes
+    results.
```
Change to:
Compares the values for a list of tuples specifying x, y points [(x1,y1), (x2, y2), ...]

stewartboyd119 replied:
Updated

acbergan commented on abaverify/processresults.py:
```diff
@@ -658,9 +678,42 @@ def write_results(results_to_write, fileName, depth=0):
-        r["computedValue"] = x_at_peak
+        r[compute_value_name] = x_at_peak
         testResults.append(r)
+    elif r[type_name] == results_tabular:
```
Add blank line before 683

stewartboyd119 replied:
Added

acbergan commented on abaverify/processresults.py:
```diff
@@ -361,7 +361,6 @@ def write_results(results_to_write, fileName, depth=0):
 
 # Load ODB
 odb = session.openOdb(name=os.path.join(os.getcwd(), jobName + '.odb'), readOnly=readOnly)
-
```
Why remove this line?

acbergan commented on readme.md:
```diff
@@ -162,6 +162,7 @@ A variety of different types of results can be extracted from the odbs and compa
 - `slope`: finds the slope of an xy data set
 - `finalValue`: finds the y value at the last increment in the xy data set
 - `x_at_peak_in_xy`: finds the x-value corresponding to the absolute peak in the y-value
+- `tabular`: given a list of tuples [(x,y)...] confirms that a corresponding point exists in requested output
```
Use same description as in processresults.py header

acbergan replied:
One more note: we are trying to adhere to the PEP-8 style. I didn't check to see this code is in compliance with PEP-8.

acbergan reviewed (changes requested):
Hey, Looks good overall!

A couple of general comments:
1. Please reorganize into two commits: one commit that replaces string literals with variables and a second commit with the tabular feature.
2. It would be nice to add a test case to the tests directory. There are some really simple tests in there currently. Let me know if you have questions about how to do this.

I made some specific comments in the diff as well.

Thanks for the PR

Commit:
*Refactored* processresults.py to eliminate use of
some string literals
==================================================================

processresults.py made heavy use of string literals, which could lead to
bugs due to typos. To address this issue many of the strings used
to index into results dicts were replaced with variables

Commit:
Updated to include "tabular" output
================================================

The tabular abaverify option allows a user to specify a list of
tuples to be confirmed to be within the tolerance of the extracted
results

Commit:
Updates according to Pull Request Review
============================================

1. Removed # --- from comments
2. renamed local variable delta to tolerance in main.py for consistancy
3. Updated readme comments and docstring describing tabular definition
4. Updated to allow tolerance to be specified as a single tuple (along
   with ability to specifiy a list of tuples)
5. There was a check within _runAssertionOnResults which checked the
   type of the "computed_val" to see if it was a tuple. In which case
   the behavior was adjusted. Changed this to check the "reference_val"
   as the output might be garbage if there's a user error

Commit:
Updated Assertion Error in _runAssertionOnResults
=======================================================

There's a check which confirms that the length of the tolerance
specified values is the same as computed vals (which was
asserted to be the same as the reference vals). This check
must be true, but in the case where a tolerance is specifed
as a single tuple a list is generated from that tuple equiv
in length to the reference vals. Therefore the user would
have specified a single tuple (which isn't the same length
as the reference values). Therefore the error message was
potentially confusing and was updated

Commit:
Added test to confirm tabular input along with bugfixes
============================================================

1. If multiple tabular verifications are requested the userData
   would throw a ValueError because ld was attempted to add
   multiple times. The integer indicating which  result object
   is being operated on is appended on ld to keep it unique
2. Bug was introduced when a single tuple for tabular tolerance was allowed.
3.

stewartboyd119 commented:
Made the updates you requested. To break up the commit into two required a rebase. Seems to work out fine. 

Long story short:

1. added a test for a tuple specified tolerance and a list of tuple specified tolerance
2. Did small cleanups you requested
3. Rebased the string literal refactoring commit to be in its own
